src/sql/workbench/api/common/extHostNotebookDocumentsAndEditors.ts

```diff
@@ -220,6 +222,43 @@ export class ExtHostNotebookDocumentsAndEditors implements ExtHostNotebookDocume
 	//#endregion
 
 	//#region Extension accessible methods
+	async createNotebookDocument(providerId: string, contents?: azdata.nb.INotebookContents): Promise<URI> {
+		let options: INotebookShowOptions = {};
+		if (contents) {
+			options.providerId = providerId;
+			options.initialContent = JSON.stringify(contents);
+		}
+		let uriComps = await this._proxy.$tryCreateNotebookDocument(options);
+		if (uriComps) {
+			let uri = URI.revive(uriComps);
+			let notebookCells = contents?.cells?.map<azdata.nb.NotebookCell>(cellContents => {
+				return {
+					contents: cellContents,
+					uri: undefined
+				};
+			});
+
+			let documentData = new ExtHostNotebookDocumentData(
+				this._proxy,
+				uri,
+				providerId,
+				false,
+				notebookCells ?? []
+			);
+			this._documents.set(uri.toString(), documentData);
+			this._onDidOpenNotebook.fire(documentData.document);
+
+			return URI.revive(uriComps);
+		} else {
+			throw new Error(`Failed to create notebook document for provider "${providerId}".`);
+		}
+	}
+
+	async openNotebookDocument(uri: vscode.Uri): Promise<azdata.nb.NotebookDocument> {
+		let docData = this._documents.get(uri.toString());
+		return docData?.document;
```

Comment:
If this can be undefined return type should be `azdata.nb.NotebookDocument`

But it seems like you should probably throw an error in that case instead

Reply:
Changed to throw an error instead.